plutus-chain-index/README_marconi.md

```diff
@@ -0,0 +1,159 @@
+
+Marconi's index component handles rollbacks by distinguishing volatile and immutable blocks. A block on the top of the chain is volatile since it is subject to some probability of being rolled back. Eventually, when the block is deeper than the security parameter, the block becomes immutable. The Marconi indexer tracks this status and keeps the volatile blocks in memory until they become immutable. Once immutable, Marconi persists the blocks to the database.
+
+### Specifying Slots to Query
+
+Through the query interface, you can specify a range of slots to query or indicate that the query is for blocks deeper than a specified number of slots. As a result, you can use simple state management for applications that work only with immutable blocks. 
+
+### Querying Volatile Data
+
+If your application requires faster response times, you can query volatile data on the head of the chain; however, you must take care to invalidate cached data when rollbacks occur. 
+
+## API Layers
+
+The core Marconi APIs are defined in Haskell in order to support simple use cases where all the components run together in a single process. 
+
+If your use cases require multiple synchronized instances for load balancing or for supporting different indexers, Marconi is designed with alternative transport layers on top of the core API that support network streaming and RPC calls. 
+
+## What Differentiates Marconi
+
+* Developed by IOG
+* Built in Haskell
+
+### DB Sync
+
+Cardano DB-sync is IOG's current indexing solution, but it uses lots of memory and days/week to sync. DB-sync takes an index everything approach to put all information into a database that can support any query at the cost of requiring a large DB server to run it. 
+
+In contrast, Marconi requires the user to customize it for the specific application so that only relevant information is streamed or stored. Marconi will be a scalable solution that allows the dApp developer to index whatever is important for their unique use case. 
+
+### Ogmios
+
+Ogmios exposes the basic node interface as a web service. This is a great option for queries that can be made directly to the node for applications that want a web service interface.
+
+### Oura
+
+Oura provides very similar functionality with an implementation in Rust and has good connectivity with cloud infrastructure like Kafka. Marconi provides similar functionality implemented in Haskell to make it easy for Haskell programmers to customize it.
+
+### Scrolls 
+
+Like Scrolls, we can be selective for what is to be indexed. Scrolls can store into multiple databases. Marconi is currently focused on local DBs like SQL lite. 
+
+### Carp 
+
+## Customizing Marconi
```

Comment:
With regards to indexing there are currently a couple of ways to customise them:

 A. A function that given events + current state emits notifications (the notification infrastructure is a little un-specified for now, so basically, I am just using it as a placeholder).

B. A function performing queries, where you can customise the following:
- The query type (this is an ADT, the point being that each user should define the types of queries that the indexer responds to).
- The point in the in-memory history where you want to run the query (this would be how i would implement querying for certain slot numbers).
- The query function's responsibility is to produce a result by merging the in-memory and on-disk data. If one only cares about on-disk data, then they could query the database directly.

C. A function that stores buffered data.
There is no connection to any storage mechanism here, so it's not like we are using SQLite because we made a choice; it simply has a nice API and is available, but changing it is *very* simple and encouraged.

With regards to customising anything else, I am not sure we can safely say anything for now, so I would remove everything that does not have to do with indexing. I am not sure what the next 4 lines describe, but I am not aware of anything in the code base that makes them true: I am not sure what the function mentioned is, Indexers have no type classes, I am not sure how you can customise the streaming component with a function (other than what the `streaming` api offers).

I also don't think that we should have any "what we want to do next" kind of things here as they may be interpreted as committing to something that we have not even properly discussed.

Reply:
This is super helpful input! I've incorporated your comments.